File: src/server/src/main/java/com/project/domain/post/Post.java
package com.project.domain.post;

import com.project.domain.BaseEntity;
import com.project.domain.category.Category;
import com.project.domain.comment.Comment;
import com.project.domain.image.PostImage;
import com.project.domain.user.User;
import com.project.utilities.Time;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;


@Entity(name = "posts")
public class Post extends BaseEntity {
    @Id
    private UUID id;
    private String title;
    private String text;
    private String previewText;
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private final List<Comment> comments;
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private final List<PostImage> postImages;

    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToOne(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(name = "posts_tags", joinColumns = @JoinColumn(name = "post_id"),
    inverseJoinColumns = @JoinColumn(name = "tag_id"))
    private final Set<Tag> tags;

    protected Post() {
        super();
        this.comments = new ArrayList<>();
        this.postImages = new ArrayList<>();
        this.tags = new HashSet<>();
    }

    public Post(String title, String text, String previewText, User user, Category category) {
        this();
        this.id = UUID.randomUUID();
        this.title = title;
        this.text = text;
        this.previewText = previewText;
        this.category = category;
        this.user = user;
    }

    public UUID getId() {
        return this.id;
    }

    public String getTitle() {
        return this.title;
    }

    public String getText() {
        return this.text;
    }

    public List<Comment> getComments() {
        return this.comments;
    }

    public List<PostImage> getPostImages() {
        return this.postImages;
    }

    public void addImagesToPost(List<PostImage> postImages) {
        this.setModifiedAt(Time.utcNow());
        this.postImages.addAll(postImages);
    }

    public String getPreviewText() {
        return this.previewText;
    }

    public User getUser() {
        return this.user;
    }

    public Category getCategory() {
        return this.category;
    }

    public Set<Tag> getTags() {
        return this.tags;
    }

    public void addTag(Tag tag) {
        this.tags.add(tag);
        tag.addPost(this);
    }

    public void removeTag(Tag tag) {
        this.tags.remove(tag);
        tag.removePost(this);
    }
}

File: src/server/src/main/java/com/project/domain/post/Post.kt
package com.project.domain.post

import com.project.domain.BaseEntity
import com.project.domain.category.Category
import com.project.domain.comment.Comment
import com.project.domain.image.PostImage
import com.project.domain.user.User
import com.project.utilities.Time.utcNow
import jakarta.persistence.CascadeType
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.JoinTable
import jakarta.persistence.ManyToMany
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import java.util.UUID

@Entity(name = "posts")
class Post protected constructor() : BaseEntity() {
    @Id
    val id: UUID = UUID.randomUUID()
    var title: String? = null
        private set
    var text: String? = null
        private set
    var previewText: String? = null
        private set

    @JvmField
    @OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    val comments: List<Comment> = mutableListOf()

    @OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    val postImages: MutableList<PostImage> = mutableListOf()

    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    var category: Category? = null
        private set

    @ManyToOne(cascade = [CascadeType.PERSIST, CascadeType.MERGE])
    @JoinColumn(name = "user_id", nullable = false)
    var user: User? = null
        private set

    @ManyToMany(cascade = [CascadeType.PERSIST, CascadeType.MERGE])
    @JoinTable(name = "posts_tags", joinColumns = [JoinColumn(name = "post_id")], inverseJoinColumns = [JoinColumn(name = "tag_id")])
    val tags: MutableSet<Tag> = mutableSetOf()

    constructor(title: String, text: String, previewText: String, user: User, category: Category) : this() {
        this.title = title
        this.text = text
        this.previewText = previewText
        this.category = category
        this.user = user
    }

    fun addImagesToPost(postImages: List<PostImage>?) {
        this.postImages.addAll(postImages!!)
        this.modifiedAt = utcNow()
    }

    fun addImageToPost(postImage: PostImage) {
        postImages.add(postImage)
        this.modifiedAt = utcNow()
    }

    fun addTag(tag: Tag) {
        tags.add(tag)
        tag.addPost(this)
        this.modifiedAt = utcNow()
    }

    fun removeTag(tag: Tag) {
        tags.remove(tag)
        tag.removePost(this)
        this.modifiedAt = utcNow()
    }
}
